src/details/ArborX_DetailsTreeTraversal.hpp

```diff
@@ -25,6 +25,42 @@ namespace ArborX
 namespace Details
 {
 
+#ifdef ARBORX_ENABLE_MPI
+// FIXME: Used only by the DistributedTree. Should be removed once proper
+// callback abstractions are in place.
+template <typename BVH, typename ReversePermutation>
+KOKKOS_FUNCTION typename BVH::bounding_volume_type const &
+getPrimitiveBoundingVolume(BVH const &bvh,
+                           ReversePermutation const &rev_permute,
+                           int leaf_index)
+{
+  int const leaf_nodes_shift = bvh.size() - 1;
+  return bvh.getBoundingVolume(
+      bvh.getNodePtr(rev_permute(leaf_index) + leaf_nodes_shift));
+}
+
+template <typename ExecutionSpace, typename BVH>
+Kokkos::View<unsigned int *, typename BVH::memory_space>
+getReversePermutation(ExecutionSpace const &exec_space, BVH const &bvh)
+{
+  auto const n = bvh.size();
+  auto const &leaf_nodes = bvh.getLeafNodes();
+
+  Kokkos::View<unsigned int *, typename BVH::memory_space> rev_permute(
+      Kokkos::ViewAllocateWithoutInitializing(
+          "ArborX::DistributedTree::query::nearest::reverse_permutation"),
+      n);
+  Kokkos::parallel_for(
+      "ArborX::DistributedTree::query::nearest::compute_reverse_permutation",
+      Kokkos::RangePolicy<ExecutionSpace>(exec_space, 0, n),
+      KOKKOS_LAMBDA(int const i) {
+        rev_permute(leaf_nodes(i).getLeafPermutationIndex()) = i;
+      });
```

Comment:
This was actually a nice idea. Mostly, because it avoids all the semantics of when and how to store user primitives, or extra arrays. It completely bypasses this question, because it is possible to figure out reverse permutation from the leave nodes, which are accessible during `query`.